File: verilog/dv/cocotb/user_proj_tests/fir_wb/fir_wb.c
// SPDX-FileCopyrightText: 2026 Mummana Jagadeesh
// SPDX-License-Identifier: Apache-2.0
#include <firmware_apis.h>

#define FIR_BASE     0x30000000
#define REG_CTRL     (*(volatile uint32_t*)(FIR_BASE + 0x00))
#define REG_OSR      (*(volatile uint32_t*)(FIR_BASE + 0x04))
#define REG_COEFF_A  (*(volatile uint32_t*)(FIR_BASE + 0x08))
#define REG_COEFF_D  (*(volatile uint32_t*)(FIR_BASE + 0x0C))
#define REG_STATUS   (*(volatile uint32_t*)(FIR_BASE + 0x10))
#define REG_PWM      (*(volatile uint32_t*)(FIR_BASE + 0x14))

void main() {
    ManagmentGpio_outputEnable();
    ManagmentGpio_write(0);
    enableHkSpi(0);

    // GPIO[8]=input, GPIO[9]=output
    GPIOs_configure(8, GPIO_MODE_USER_STD_INPUT_NOPULL);
    GPIOs_configure(9, GPIO_MODE_USER_STD_OUT_MONITORED);
    GPIOs_loadConfigs();
    User_enableIF();

    // Signal start
    ManagmentGpio_write(1);

    // Soft reset
    REG_CTRL = 0x08;
    for (volatile int i = 0; i < 100; i++);
    REG_CTRL = 0x00;

    // Set OSR=16
    REG_OSR = 16;

    // Load moving average coefficients (h[k]=0x1000 = 0.125 in Q1.15)
    for (int i = 0; i < 8; i++) {
        REG_COEFF_A = i;
        REG_COEFF_D = 0x1000;
    }

    // Enable with LFSR test source
    REG_CTRL = 0x11; // enable=1, use_lfsr=1

    // Wait for pipeline to fill
    for (volatile int i = 0; i < 5000; i++);

    // Signal done — cocotb monitors this
    ManagmentGpio_write(0);

    // Test bypass mode: direct PWM = 0x80 (50%)
    REG_PWM  = 0x80;
    REG_CTRL = 0x03; // enable=1, bypass_fir=1

    for (volatile int i = 0; i < 1000; i++);

    // Signal complete
    ManagmentGpio_write(1);

    while(1);
}

File: verilog/dv/cocotb/user_proj_tests/fir_wb/fir_wb.py
# SPDX-FileCopyrightText: 2026 Mummana Jagadeesh
# SPDX-License-Identifier: Apache-2.0
from caravel_cocotb.caravel_interfaces import test_configure
from caravel_cocotb.caravel_interfaces import report_test
import cocotb
from cocotb.triggers import ClockCycles, Timer

@cocotb.test()
@report_test
async def fir_wb(dut):
    caravelEnv = await test_configure(dut, timeout_cycles=500000)
    cocotb.log.info("[TEST] Start fir_wb test")

    await caravelEnv.release_csb()

    # Wait for firmware to signal start (mgmt_gpio=1)
    await caravelEnv.wait_mgmt_gpio(1)
    cocotb.log.info("[TEST] Firmware started - FIR pipeline initializing")

    # Wait for pipeline done signal (mgmt_gpio=0)
    await caravelEnv.wait_mgmt_gpio(0)
    cocotb.log.info("[TEST] Pipeline filled")

    # Wait a bit then check bypass mode signal
    await caravelEnv.wait_mgmt_gpio(1)
    cocotb.log.info("[TEST] Bypass mode active - fir_wb PASSED")
